// native-base/lib/commonjs/components/primitives/Spinner/variants/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Default = Default;

var _react = _interopRequireDefault(require("react"));

var _Icon = require("../../../primitives/Icon");

var _reactNativeSvg = require("react-native-svg");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Default(props) {
  return /*#__PURE__*/_react.default.createElement(_Icon.Icon, {
    color: props.color,
    viewBox: "0 0 64 64",
    size: props.size
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.G, {
    id: "Group_182",
    "data-name": "Group 182",
    transform: "translate(-258 -488)"
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.G, {
    id: "Ellipse_8",
    "data-name": "Ellipse 8",
    transform: "translate(258 488)",
    fill: "none",
    stroke: "#e8edfb",
    strokeWidth: "6"
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Circle, {
    cx: "32",
    cy: "32",
    r: "32",
    stroke: "none"
  }), /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Circle, {
    cx: "32",
    cy: "32",
    r: "29",
    fill: "none"
  })), /*#__PURE__*/_react.default.createElement(_reactNativeSvg.G, {
    id: "Ellipse_10",
    "data-name": "Ellipse 10",
    transform: "translate(258 488)",
    fill: "none",
    stroke: props.color,
    strokeLinecap: "square",
    strokeWidth: "6",
    strokeDasharray: "90 10000"
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Circle, {
    cx: "32",
    cy: "32",
    r: "32",
    stroke: "none"
  }), /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Circle, {
    cx: "32",
    cy: "32",
    r: "29",
    fill: "none"
  }))));
}
