// native-base/lib/commonjs/components/composites/Typeahead/useTypeahead/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ControlledPropUpdatedSelectedItem = exports.FunctionReset = exports.FunctionSetInputValue = exports.FunctionSelectItem = exports.FunctionCloseMenu = exports.FunctionOpenMenu = exports.FunctionToggleMenu = exports.ToggleButtonClick = exports.ItemClick = exports.InputBlur = exports.InputChange = void 0;
const InputChange = '__input_change__';
exports.InputChange = InputChange;
const InputBlur = '__input_blur__';
exports.InputBlur = InputBlur;
const ItemClick = '__item_click__';
exports.ItemClick = ItemClick;
const ToggleButtonClick = '__togglebutton_click__';
exports.ToggleButtonClick = ToggleButtonClick;
const FunctionToggleMenu = '__function_toggle_menu__';
exports.FunctionToggleMenu = FunctionToggleMenu;
const FunctionOpenMenu = '__function_open_menu__';
exports.FunctionOpenMenu = FunctionOpenMenu;
const FunctionCloseMenu = '__function_close_menu__';
exports.FunctionCloseMenu = FunctionCloseMenu;
const FunctionSelectItem = '__function_select_item__';
exports.FunctionSelectItem = FunctionSelectItem;
const FunctionSetInputValue = '__function_set_input_value__';
exports.FunctionSetInputValue = FunctionSetInputValue;
const FunctionReset = '__function_reset__';
exports.FunctionReset = FunctionReset;
const ControlledPropUpdatedSelectedItem = '__controlled_prop_updated_selected_item__';
exports.ControlledPropUpdatedSelectedItem = ControlledPropUpdatedSelectedItem;
