// native-base/lib/commonjs/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getSpacedChildren", {
  enumerable: true,
  get: function () {
    return _getSpacedChildren.default;
  }
});
Object.defineProperty(exports, "getAbsoluteChildren", {
  enumerable: true,
  get: function () {
    return _getAbsoluteChildren.default;
  }
});
Object.defineProperty(exports, "getAttachedChildren", {
  enumerable: true,
  get: function () {
    return _getAttachedChildren.default;
  }
});
Object.defineProperty(exports, "addTextAndPropsToStrings", {
  enumerable: true,
  get: function () {
    return _addTextAndPropsToStrings.addTextAndPropsToStrings;
  }
});
Object.defineProperty(exports, "canUseDom", {
  enumerable: true,
  get: function () {
    return _canUseDom.canUseDom;
  }
});
Object.defineProperty(exports, "mergeRefs", {
  enumerable: true,
  get: function () {
    return _mergeRefs.mergeRefs;
  }
});
Object.defineProperty(exports, "composeEventHandlers", {
  enumerable: true,
  get: function () {
    return _mergeRefs.composeEventHandlers;
  }
});
Object.defineProperty(exports, "wrapStringChild", {
  enumerable: true,
  get: function () {
    return _wrapStringChild.wrapStringChild;
  }
});
Object.defineProperty(exports, "combineContextAndProps", {
  enumerable: true,
  get: function () {
    return _combineContextAndProps.combineContextAndProps;
  }
});
Object.defineProperty(exports, "ariaAttr", {
  enumerable: true,
  get: function () {
    return _accessibilityUtils.ariaAttr;
  }
});
Object.defineProperty(exports, "createContext", {
  enumerable: true,
  get: function () {
    return _createContext.createContext;
  }
});
Object.defineProperty(exports, "useKeyboardBottomInset", {
  enumerable: true,
  get: function () {
    return _useKeyboardBottomInset.useKeyboardBottomInset;
  }
});
Object.defineProperty(exports, "isEmptyObj", {
  enumerable: true,
  get: function () {
    return _isEmptyObj.isEmptyObj;
  }
});
Object.defineProperty(exports, "resolveStackStyleInput", {
  enumerable: true,
  get: function () {
    return _resolveStackStyleInput.resolveStackStyleInput;
  }
});
Object.defineProperty(exports, "getStyleElement", {
  enumerable: true,
  get: function () {
    return _useResponsiveQuery.getStyleElement;
  }
});

var _getSpacedChildren = _interopRequireDefault(require("./getSpacedChildren"));

var _getAbsoluteChildren = _interopRequireDefault(require("./getAbsoluteChildren"));

var _getAttachedChildren = _interopRequireDefault(require("./getAttachedChildren"));

var _addTextAndPropsToStrings = require("./addTextAndPropsToStrings");

var _canUseDom = require("./canUseDom");

var _mergeRefs = require("./mergeRefs");

var _wrapStringChild = require("./wrapStringChild");

var _combineContextAndProps = require("./combineContextAndProps");

var _accessibilityUtils = require("./accessibilityUtils");

var _createContext = require("./createContext");

var _useKeyboardBottomInset = require("./useKeyboardBottomInset");

var _isEmptyObj = require("./isEmptyObj");

var _resolveStackStyleInput = require("./resolveStackStyleInput");

var _useResponsiveQuery = require("./useResponsiveQuery/useResponsiveQuery");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
